### PR TITLE
🚀 Remove tagname allowlist for loading indicator

### LINKED ISSUE #33543: Intent-to-Remove: loader allowlist
We show loading indicators only for a select group of elements:

https://github.com/ampproject/amphtml/blob/246b09c4f8c80a621289d3b9dcb67436af94a291/src/layout.js#L96-L118

This list gets shipped as part of [`v0.js`](https://cdn.ampproject.org/v0.js). Introducing a new component with a loader **will then increase the size of `v0.js`** roughly by the size of its name, e.g. `'AMP-TIKTOK': !0`. 

## This proposes removing the allowlist by tagname

-   At one point, loaders were changed so that they would [only show up after 600ms of loading time](https://github.com/ampproject/amphtml/blob/212f4fe289e351f22090b10c5eb7508375c97a2e/extensions/amp-loader/0.1/amp-loader.js#L31-L33). **Loaders appear less frequently than when the list was introduced.**

-   The list is not widely known among contributors, so it's likely very out of date, and would further increase the size of v0.js if updated. **Switching to a simpler ruleset would prevent v0.js from growing further**, and would ensure that components get the same treatment when new.

### Replacement rule

Loaders will now be rendered any time a component's layout is **not** `container` **nor** `nodisplay`.

🔵  **includes**:

- `amp-img`
- `amp-video`
- `amp-iframe`
- `amp-list`
- (and more, see validation rules)

 ⛔️ **excludes**:

- `amp-story`
- `amp-sidebar`
- `amp-accordion`
- `amp-fit-text`
- (and more, see validation rules)

---

cc @ampproject/wg-components 
cc @ampproject/wg-performance

### PR DESCRIPTION
**Pull request by alanorozco**
Fixes #33543

Notable:

- `dist/v0.js`: Δ -0.18KB
- `dist/v0.mjs`: Δ -0.20KB

### CODE REVIEW

**Comment by alanorozco** [comment]
@caroqliu Since `amp-list` does `toggleLoading` explicitly when `layout=container`, this change should not change its behavior, correct? (See unchanged `layout=container` test.)

File: build-system/test-configs/forbidden-terms.js [hunk 1]
```diff
@@ -764,7 +764,6 @@ const forbiddenTermsGlobal = {
       'test/unit/test-intersection-observer-polyfill.js',
       'test/unit/test-json.js',
       'test/unit/test-layout-rect.js',
-      'test/unit/test-layout.js',
```
Comment (alanorozco) [inline build-system/test-configs/forbidden-terms.js:767]: Not related to this change: the regex for this term is incorrect since it misses `describe` methods after a newline:

```js
describe
  .configure()
  ...
```

This file **should** match. However, if I fix the pattern I'd have to allow a bunch of unrelated files so I'd rather fix that separately.

**Comment by samouri** [comment]
> @caroqliu Since amp-list does toggleLoading explicitly when layout=container, this change should not change its behavior, correct?

Theoretically it would only change in one condition. If the `<amp-list>` extension JS took >600ms to load/execute.
Previously, v0 would have turned on the loader for that case.

**Comment by samouri** [comment]
> @caroqliu Since amp-list does toggleLoading explicitly when layout=container, this change should not change its behavior, correct?

Theoretically it would only change in one condition. If the `<amp-list>` extension JS took >600ms to load/execute.
Previously, v0 would have turned on the loader for that case.

**Comment by alanorozco** [comment]
> If the <amp-list> extension JS took >600ms to load/execute. Previously, v0 would have turned on the loader for that case.

Noting that this condition only affects `<amp-list layout="container">`, which is experimental and email only:

https://github.com/ampproject/amphtml/blob/d53c7be8e8f5374fb826d2ed28f7565cd6e899e8/extensions/amp-list/validator-amp-list.protoascii#L233

File: extensions/amp-loader/0.1/amp-loader.js [hunk 2]
```diff
@@ -47,6 +46,39 @@ const LOADER_BACKGROUND_TAGS = {
...
+ * reasons, so they are listed individually.
+ * @private @enum {boolean}
+ */
+const VIDEO_PLAYER_TAGS_LEGACY = {
+  'AMP-BRIGHTCOVE': true,
+  'AMP-DAILYMOTION': true,
+  'AMP-YOUTUBE': true,
+  'AMP-VIMEO': true,
+  'AMP-IMA-VIDEO': true,
+};
```
Comment (alanorozco) [inline extensions/amp-loader/0.1/amp-loader.js:67]: I'm not a fan of the `amp-loader` extension containing this list nor `LOADER_BACKGROUND_TAGS` above.

https://github.com/ampproject/amphtml/blob/58eddda4a07ab8bab2112071845c87c2986953d5/extensions/amp-loader/0.1/amp-loader.js#L34-L47

However, these lists are used in a more subtle manner than the logic to display loaders in `v0`, so I think it's best to optimize those separately.

File: extensions/amp-loader/0.1/amp-loader.js [hunk 3]
```diff
@@ -47,6 +46,38 @@ const LOADER_BACKGROUND_TAGS = {
...
+ * A few components don't follow the previous convention for historical
+ * reasons, so they are listed individually.
+ * @private @enum {boolean}
+ */
+const VIDEO_PLAYER_TAGS_LEGACY = {
+  'AMP-BRIGHTCOVE': true,
+  'AMP-DAILYMOTION': true,
+  'AMP-YOUTUBE': true,
+  'AMP-VIMEO': true,
+};
```
Comment (alanorozco) [inline extensions/amp-loader/0.1/amp-loader.js:66]: I'm not a fan of the `amp-loader` extension containing this list nor `LOADER_BACKGROUND_TAGS` above.

https://github.com/ampproject/amphtml/blob/58eddda4a07ab8bab2112071845c87c2986953d5/extensions/amp-loader/0.1/amp-loader.js#L34-L47

However, these lists are used in a more subtle manner than the logic to display loaders in `v0`, so I think it's best to deal with those separately. I did however take the liberty to remove `AMP-IMA-VIDEO`:

```diff
- const VIDEO_PLAYER_TAG_PATTERN = /^amp\-(video|.+player)/;
+ const VIDEO_PLAYER_TAG_PATTERN = /\-(video|.+player)/;

  const VIDEO_PLAYER_TAGS_LEGACY = {
    ...
-   'AMP-IMA-VIDEO': true,
  };
```

**Comment by caroqliu** [comment]
> @caroqliu Since `amp-list` does `toggleLoading` explicitly when changed to `layout=container`, this PR should not affect its behavior. Is that correct?
> 
> See unchanged `layout=container` test:
> 
> https://github.com/ampproject/amphtml/blob/0a96130ef7dacb46d3912945fbd99becdbd5cfe1/test/unit/test-custom-element.js#L2201-L2205

Yup, I'm seeing the loader still work from manually viewing `examples/amp-list-layout-container.amp.html` on your branch.

**Comment by alanorozco** [comment]
> the property you should be checking for R1 components is `usesLoading`:

That's in the component implementation class, right? If so, we can't use that since `isLoadingAllowed` is sync and called by core.

**Comment by samouri** [comment]
> > the property you should be checking for R1 components is `usesLoading`:
> 
> That's in the component implementation class, right? If so, we can't use that since `isLoadingAllowed` is sync and called by core.

In that case I think we can remove the loadable condition.

File: extensions/amp-loader/0.1/amp-loader.js [hunk 4]
```diff
@@ -47,6 +46,38 @@ const LOADER_BACKGROUND_TAGS = {
   'AMP-GOOGLE-DOCUMENT-EMBED': true,
 };
 
+/**
+ * All video player components must either have a) "video" or b) "player" in
+ * their name.
+ * @private @const {!RegExp}
+ */
+const VIDEO_PLAYER_TAG_PATTERN = /\-(video|player)/i;
```
Comment (newmuis) [inline extensions/amp-loader/0.1/amp-loader.js:54]: FWIW, this regex also matches `amp-story-player`, which is not a video player.

/cc @Enriqe @gmajoulet

Comment (alanorozco) [reply extensions/amp-loader/0.1/amp-loader.js:54]: This has historically been the case. Its effect is that it draws a specific background color behind the placeholder.

Comment (newmuis) [reply extensions/amp-loader/0.1/amp-loader.js:54]: Makes sense, but @Enriqe @gmajoulet are we aware of this?

> Its effect is that it draws a specific background color behind the placeholder.

I think my concern is that if we don't actually desire this semantically (regardless of what the specific effects are today), we'll get "surprised" when loaders have a requirements change and they "accidentally" get applied to `amp-story-player` as well.

Of course, this is totally fine if we just actually want this to apply to `amp-story-player` too, I just don't know.

Comment (gmajoulet) [reply extensions/amp-loader/0.1/amp-loader.js:54]: Can we update the regex and add a negative lookahead to exclude `story-player`?

Comment (alanorozco) [reply extensions/amp-loader/0.1/amp-loader.js:54]: Done, but please take note:

- You were seeing a loader on `amp-story-player` previously because it would match this regex, it was not listed.
- You will **still** see a loader after excluding it from this regex, but you won't see the background for which there's a list.

- After this change, you can't opt-out of loaders: if you're sized (e.g. not `layout=container`) and you resolve an async `layoutCallback` after `600ms` you load, that's that.
- Loader logic is likely to be simplified

Comment (gmajoulet) [reply extensions/amp-loader/0.1/amp-loader.js]: @Enriqe can you take a look?

Comment (Enriqe) [reply extensions/amp-loader/0.1/amp-loader.js]: Sorry, this got lost in my notifications.

Thanks @alanorozco for the context and additional information. For context, the player has its own loader that will eventually listen to a story-specific signal to stop displaying (see https://github.com/ampproject/amphtml/issues/32436). I think we can just use `this.toggleLoading(false)` like you mentioned to disable the default AMP loader.

For now, having this updated regex that excludes the `amp-story-player` from having the loader background color SGTM. 

> (Possibly, later in the future) Loader logic is likely to be simplified 

Could you please make sure to cc `@ampproject/wg-stories` so that we are aware of these future changes too? Thanks!
